src/core/model/Monitor.cpp

```diff
@@ -0,0 +1,57 @@
+#include "Monitor.h"
+
+/*
+ * Credit to Mike Vine on StackOverflow.
+ * https://stackoverflow.com/a/48408987
+ */
+
+template <class T>
+template<typename ...Args>
+Monitor<T>::Monitor(Args&&... args) : model(std::forward<Args>(args)...) {}
+
+
+template <class T>
+Monitor<T>::LockedMonitor::LockedMonitor(Monitor* monitor) : mon(monitor), lock(monitor->mutex) {}
+
+template <class T>
+T* Monitor<T>::LockedMonitor::operator->() { return &mon->model;}
+
+template <class T>
+void Monitor<T>::LockedMonitor::ReplaceModel(T model) {
+    mon->model = model;
+}
+
+
+template <class T>
+Monitor<T>::TryLockedMonitor::TryLockedMonitor(Monitor* monitor) {
+    this->lock = std::unique_lock<std::mutex>(monitor->mutex, std::defer_lock);
+    this->lockAcquired = lock.try_lock();
+    if (this->lockAcquired) {
+        this->mon = monitor;
+    }
+}
+
+template <class T>
+T* Monitor<T>::TryLockedMonitor::operator->() {
+    assert(lockAcquired);
+    return &mon->model;
+};
+
+template <class T>
+void Monitor<T>::TryLockedMonitor::ReplaceModel(T model) {
+    assert(lockAcquired);
+    mon->model = model;
+}
```

Comment:
I think, overloading operator== is what we want. Also, for the Monitor class itself.